lib/vector-config/src/schema.rs

```diff
@@ -479,30 +461,32 @@ pub fn generate_internal_tagged_variant_schema(
 
 pub fn generate_root_schema<T>() -> Result<RootSchema, GenerateError>
 where
-    T: Configurable,
+    T: Configurable + 'static,
 {
     // Set env variable to enable generating all schemas, including platform-specific ones.
     std::env::set_var("VECTOR_GENERATE_SCHEMA", "true");
 
     let schema_gen = RefCell::new(SchemaSettings::new().into_generator());
 
-    let schema = get_or_generate_schema::<T>(&schema_gen, Some(T::metadata()))?;
+    let schema = get_or_generate_schema(&T::make_ref(), &schema_gen, Some(T::metadata()))?;
+
     Ok(schema_gen.into_inner().into_root_schema(schema))
 }
 
-pub fn get_or_generate_schema<T>(
+pub fn get_or_generate_schema(
+    config: &ConfigurableRef,
     gen: &RefCell<SchemaGenerator>,
     overrides: Option<Metadata>,
-) -> Result<SchemaObject, GenerateError>
-where
-    T: Configurable,
-{
-    let (mut schema, metadata) = match T::referenceable_name() {
+) -> Result<SchemaObject, GenerateError> {
+    let metadata = config.make_metadata();
+    let (mut schema, metadata) = match &config.referenceable_name() {
         // When `T` has a referenceable name, try looking it up in the schema generator's definition
         // list, and if it exists, create a schema reference to it. Otherwise, generate it and
         // backfill it in the schema generator.
         Some(name) => {
-            if !gen.borrow().definitions().contains_key(name) {
+            // Creating a string just to borrow it is ugly, but that's required by the types in the
+            // schema generator definitions.
+            if !gen.borrow().definitions().contains_key(&name.to_string()) {
                 // In order to avoid infinite recursion, we copy the approach that `schemars` takes and
```

Comment:
`name` should be `&'static str`, and `IndexMap::contains_key` should function equivalently to `HashMap::contains_key`, where if `K` is `String`, we should be able to use `&str` as a key for lookups.. so I feel like I'm missing something here. 🤔

Reply:
This is the error I get just using `name` here. I couldn't figure another formulation that works.
```
error[E0277]: the trait bound `std::string::String: Borrow<&str>` is not satisfied
   --> lib/vector-config/src/schema.rs:487:57
    |
487 |             if !gen.borrow().definitions().contains_key(name) {
    |                                            ------------ ^^^^ the trait `Borrow<&str>` is not implemented for `std::string::String`
    |                                            |
    |                                            required by a bound introduced by this call
    |
    = help: the trait `Borrow<str>` is implemented for `std::string::String`
    = note: required for `&str` to implement `Equivalent<std::string::String>`
note: required by a bound in `IndexMap::<K, V, S>::contains_key`
   --> /home/bruce.guenter/.cargo/registry/src/github.com-1ecc6299db9ec823/indexmap-1.9.2/src/map.rs:400:19
    |
400 |         Q: Hash + Equivalent<K>,
    |                   ^^^^^^^^^^^^^ required by this bound in `IndexMap::<K, V, S>::contains_key`

For more information about this error, try `rustc --explain E0277`.
```

Reply:
Ohhh. Drop that extra reference on line 482.

Reply:
Oh, duh, missed the obvious there. Fixed in eddc6a19f